hummingbot/connector/utils.py

```diff
@@ -51,3 +54,40 @@ def split_hb_trading_pair(trading_pair: str) -> Tuple[str, str]:
 def combine_to_hb_trading_pair(base: str, quote: str) -> str:
     trading_pair = f"{base}-{quote}"
     return trading_pair
+
+
+def get_new_client_order_id(
+    is_buy: bool, trading_pair: str, hbot_order_id_prefix: str = "", max_id_len: Optional[int] = None
+) -> str:
+    """
+    Creates a client order id for a new order
+    :param is_buy: True if the order is a buy order, False otherwise
+    :param trading_pair: the trading pair the order will be operating with
+    :param hbot_order_id_prefix: The hummingbot-specific identifier for the given exchange
+    :param max_id_len: The maximum length of the ID string.
+    :return: an identifier for the new order to be used in the client
+    """
+    side = "B" if is_buy else "S"
+    symbols = split_hb_trading_pair(trading_pair)
+    base = symbols[0].upper()
+    quote = symbols[1].upper()
+    base_str = f"{base[0]}{base[-1]}"
+    quote_str = f"{quote[0]}{quote[-1]}"
+    client_instance_id = hex(abs(hash(f"{socket.gethostname()}{os.getpid()}")))[2:6]
+    client_order_id = ""
+    if hbot_order_id_prefix:
+        client_order_id = f"{hbot_order_id_prefix}"
+    if max_id_len is None:
+        client_order_id = f"{client_order_id}{side}{base_str}{quote_str}{client_instance_id}{get_tracking_nonce_short()}"
+    else:
+        client_order_id = f"{client_order_id}{side}{base_str}{quote_str}{client_instance_id}"
+        remaining = max_id_len - len(client_order_id)
+        if remaining < nonce_multiplier_power + 8:  # 8 digits in the seconds will give us ~1157 days of unique IDs
+            raise NotImplementedError(
+                f"The order ID exceeds the max length of {max_id_len}, but truncating {remaining}"
+                f" digits from the tracking nonce will cause duplicates in"
+                f" {(remaining - nonce_multiplier_power) / (60 * 60 * 24)} days."
+            )
+        tracking_nonce = str(get_tracking_nonce_short())[-remaining:]
+        client_order_id = f"{client_order_id}{tracking_nonce}"
```

Comment:
I would change this implementation to make sure that we always use as many characters from the instance id as possible (instead of fixing it in 4). I believe the resulting client_client_order_id should be:
```
instance_id_length = max_id_len - 5 - 10 - len(hbot_order_id_prefix) # Five is the length required for side, base_str and quote_str
# Ten is the size of tracking nonce if we use the existing nonce low res and we use only 10 digits (81 days)
instance_id = client_instance_id[0:instance_id_length] if instance_id_length > 0 else ""
f"{hbot_order_id_prefix}{side}{base_str}{quote_str}{instance_id}{str(get_tracking_nonce_low_res)[-10:]}"
```

Also I would not make this function fail. If even without the instance id the resulting order id is too long the connector should receive an error from the exchange. But if the exchange does not fail, we should not make the client fail.

Reply:
Addressed as per our discussion